File: ruby/easy/count_and_say.rb
# @param {Integer} n
# @return {String}
def count_and_say(n)
    return count_and_say_recursive(1, n, "1")
end

def count_and_say_recursive(index, n, str)
    return str if (index == n)
    i, count = 0, 1
    res = ""
    while i < str.length-1
        if str[i] == str[i+1]
            count += 1
        else
            res += "#{count}#{str[i]}"
            count = 1
        end
        i += 1
    end
    res += "#{count}#{str[i]}"
    return count_and_say_recursive(index+1, n, res)
end

File: javascript/easy/countAndSay.js
/**
 * @param {number} n
 * @return {string}
 */
var countAndSay = function(n) {
    return countAndSayRecursive(1, n, "1");
};

var countAndSayRecursive = function(i, n, str) {
    if(i == n) return str;
    let res = "";
    for(let j = 0, count = 1; j < str.length; j++) {
        if(j < str.length-1 && str[j] == str[j+1]) {
            count++;
        } else {
            res += count + str[j];
            count = 1;
        }
    }
    return countAndSayRecursive(i+1, n, res);
};
